packages/core/src/components/slider/multiSlider.tsx

```diff
@@ -0,0 +1,272 @@
+        if (foundHandle) {
+            foundHandle.beginHandleTouchMovement(event);
+        }
+    }
+
+    private renderTrackFill(index: number, start: ISliderHandleProps, end: ISliderHandleProps, intent: Intent) {
+        const { tickSizeRatio } = this.state;
+        const startValue = start.value;
+        const endValue = end.value;
+
+        let startOffsetRatio = this.getOffsetRatio(startValue, tickSizeRatio);
+        let endOffsetRatio = this.getOffsetRatio(endValue, tickSizeRatio);
+
+        if (startOffsetRatio > endOffsetRatio) {
+            const temp = endOffsetRatio;
+            endOffsetRatio = startOffsetRatio;
+            startOffsetRatio = temp;
+        }
+
+        const startOffset = formatPercentage(startOffsetRatio);
+        const endOffset = formatPercentage(1 - endOffsetRatio);
+
+        const style: React.CSSProperties = this.props.vertical
+            ? { bottom: startOffset, top: endOffset, left: 0 }
+            : { left: startOffset, right: endOffset, top: 0 };
+
+        const classes = classNames(Classes.SLIDER_PROGRESS, intentClass(intent), {
+            [Classes.START]: start.type === SliderHandleType.START,
+            [Classes.END]: end.type === SliderHandleType.END,
+            [`${Classes.SLIDER_PROGRESS}-empty`]: intent === Intent.NONE,
+        });
+
+        return <div key={`track-${index}`} className={classes} style={style} />;
+    }
+
+    private getTrackStops(): ISliderHandleProps[] {
+        const maybeStops = React.Children.map(
+            this.props.children,
+            child => (Utils.isElementOfType(child, SliderTrackStop) ? child.props : null),
+        );
+        return maybeStops != null ? maybeStops : [];
+    }
+
+    private getOffsetRatio(value: number, tickSizeRatio: number) {
+        return Utils.clamp((value - this.props.min) * tickSizeRatio, 0, 1);
+    }
+
+    private nearestHandleForValue(handles: Handle[], getOffset: (handle: Handle) => number): Handle | undefined {
+        return argMin(handles, handle => {
+            const offset = getOffset(handle);
+            const offsetValue = handle.clientToValue(offset);
+            const handleValue = handle.props.value!;
+            return Math.abs(offsetValue - handleValue);
+        });
+    }
+
+    private addHandleRef = (ref: Handle) => {
+        if (ref != null) {
+            this.handles.push(ref);
+        }
+    };
+
+    private getHandlerForIndex = (index: number, callback?: (values: number[]) => void) => {
+        return (newValue: number) => {
+            Utils.safeInvoke(callback, this.getNewHandleValues(newValue, index));
+        };
+    };
+
+    private getNewHandleValues(newValue: number, oldIndex: number) {
+        const oldValues = this.sortedHandleProps.map(handle => handle.value);
+        const start = oldValues.slice(0, oldIndex);
+        const end = oldValues.slice(oldIndex + 1);
+        const newValues = [...start, newValue, ...end];
```

Comment:
wait why not use `array.splice()`?

but actually, does insertion index really matter if you're just gonna sort again?

Reply:
Not really, we just need to replace the value at `oldIndex` with `newValue`, refactored it in a nice way I think.